// ng-http-loader/bundles/ng-http-loader.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('rxjs'), require('rxjs/operators'), require('@angular/common/http'), require('@angular/common')) :
    typeof define === 'function' && define.amd ? define('ng-http-loader', ['exports', '@angular/core', 'rxjs', 'rxjs/operators', '@angular/common/http', '@angular/common'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory(global['ng-http-loader'] = {}, global.ng.core, global.rxjs, global.rxjs.operators, global.ng.common.http, global.ng.common));
}(this, (function (exports, i0, rxjs, operators, http, i11) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () {
                            return e[k];
                        }
                    });
                }
            });
        }
        n['default'] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);
    var i11__namespace = /*#__PURE__*/_interopNamespace(i11);

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (Object.prototype.hasOwnProperty.call(b, p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }
    var __assign = function () {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s)
                    if (Object.prototype.hasOwnProperty.call(s, p))
                        t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };
    function __rest(s, e) {
        var t = {};
        for (var p in s)
            if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
                t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }
    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
            r = Reflect.decorate(decorators, target, key, desc);
        else
            for (var i = decorators.length - 1; i >= 0; i--)
                if (d = decorators[i])
                    r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }
    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); };
    }
    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function")
            return Reflect.metadata(metadataKey, metadataValue);
    }
    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try {
                step(generator.next(value));
            }
            catch (e) {
                reject(e);
            } }
            function rejected(value) { try {
                step(generator["throw"](value));
            }
            catch (e) {
                reject(e);
            } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }
    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function () { if (t[0] & 1)
                throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function () { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f)
                throw new TypeError("Generator is already executing.");
            while (_)
                try {
                    if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done)
                        return t;
                    if (y = 0, t)
                        op = [op[0] & 2, t.value];
                    switch (op[0]) {
                        case 0:
                        case 1:
                            t = op;
                            break;
                        case 4:
                            _.label++;
                            return { value: op[1], done: false };
                        case 5:
                            _.label++;
                            y = op[1];
                            op = [0];
                            continue;
                        case 7:
                            op = _.ops.pop();
                            _.trys.pop();
                            continue;
                        default:
                            if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                                _ = 0;
                                continue;
                            }
                            if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) {
                                _.label = op[1];
                                break;
                            }
                            if (op[0] === 6 && _.label < t[1]) {
                                _.label = t[1];
                                t = op;
                                break;
                            }
                            if (t && _.label < t[2]) {
                                _.label = t[2];
                                _.ops.push(op);
                                break;
                            }
                            if (t[2])
                                _.ops.pop();
                            _.trys.pop();
                            continue;
                    }
                    op = body.call(thisArg, _);
                }
                catch (e) {
                    op = [6, e];
                    y = 0;
                }
                finally {
                    f = t = 0;
                }
            if (op[0] & 5)
                throw op[1];
            return { value: op[0] ? op[1] : void 0, done: true };
        }
    }
    var __createBinding = Object.create ? (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        Object.defineProperty(o, k2, { enumerable: true, get: function () { return m[k]; } });
    }) : (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        o[k2] = m[k];
    });
    function __exportStar(m, o) {
        for (var p in m)
            if (p !== "default" && !Object.prototype.hasOwnProperty.call(o, p))
                __createBinding(o, m, p);
    }
    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m)
            return m.call(o);
        if (o && typeof o.length === "number")
            return {
                next: function () {
                    if (o && i >= o.length)
                        o = void 0;
                    return { value: o && o[i++], done: !o };
                }
            };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }
    /** @deprecated */
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }
    /** @deprecated */
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++)
            s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }
    function __spreadArray(to, from) {
        for (var i = 0, il = from.length, j = to.length; i < il; i++, j++)
            to[j] = from[i];
        return to;
    }
    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }
    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n])
            i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try {
            step(g[n](v));
        }
        catch (e) {
            settle(q[0][3], e);
        } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length)
            resume(q[0][0], q[0][1]); }
    }
    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }
    function __asyncValues(o) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function (v) { resolve({ value: v, done: d }); }, reject); }
    }
    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) {
            Object.defineProperty(cooked, "raw", { value: raw });
        }
        else {
            cooked.raw = raw;
        }
        return cooked;
    }
    ;
    var __setModuleDefault = Object.create ? (function (o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function (o, v) {
        o["default"] = v;
    };
    function __importStar(mod) {
        if (mod && mod.__esModule)
            return mod;
        var result = {};
        if (mod != null)
            for (var k in mod)
                if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k))
                    __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }
    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }
    function __classPrivateFieldGet(receiver, state, kind, f) {
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a getter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot read private member from an object whose class did not declare it");
        return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
    }
    function __classPrivateFieldSet(receiver, state, value, kind, f) {
        if (kind === "m")
            throw new TypeError("Private method is not writable");
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a setter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot write private member to an object whose class did not declare it");
        return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
    }

    /*
     * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
     * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
     * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
     * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
     * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
     * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
     */
    var AbstractLoaderDirective = /** @class */ (function () {
        function AbstractLoaderDirective() {
        }
        return AbstractLoaderDirective;
    }());
    AbstractLoaderDirective.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.0.0", ngImport: i0__namespace, type: AbstractLoaderDirective, deps: [], target: i0__namespace.ɵɵFactoryTarget.Directive });
    AbstractLoaderDirective.ɵdir = i0__namespace.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.0.0", type: AbstractLoaderDirective, inputs: { backgroundColor: "backgroundColor" }, ngImport: i0__namespace });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.0.0", ngImport: i0__namespace, type: AbstractLoaderDirective, decorators: [{
                type: i0.Directive
            }], propDecorators: { backgroundColor: [{
                    type: i0.Input
                }] } });

    var SkChasingDotsComponent = /** @class */ (function (_super) {
        __extends(SkChasingDotsComponent, _super);
        function SkChasingDotsComponent() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        return SkChasingDotsComponent;
    }(AbstractLoaderDirective));
    SkChasingDotsComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.0.0", ngImport: i0__namespace, type: SkChasingDotsComponent, deps: null, target: i0__namespace.ɵɵFactoryTarget.Component });
    SkChasingDotsComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.0.0", type: SkChasingDotsComponent, selector: "sk-chasing-dots", usesInheritance: true, ngImport: i0__namespace, template: "<!--\nCopyright (c) 2015 Tobias Ahlin\n\nPermission is hereby granted, free of charge, to any person obtaining a copy of\nthis software and associated documentation files (the \"Software\"), to deal in\nthe Software without restriction, including without limitation the rights to\nuse, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of\nthe Software, and to permit persons to whom the Software is furnished to do so,\nsubject to the following conditions:\n\nThe above copyright notice and this permission notice shall be included in all\ncopies or substantial portions of the Software.\n\nTHE SOFTWARE IS PROVIDED \"AS IS\", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR\nIMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS\nFOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR\nCOPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER\nIN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN\nCONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.\n-->\n\n<div class=\"sk-chasing-dots\" [class.colored]=\"!backgroundColor\">\n    <div class=\"sk-child sk-dot1\" [style.background-color]='backgroundColor'></div>\n    <div class=\"sk-child sk-dot2\" [style.background-color]='backgroundColor'></div>\n</div>\n", styles: [".sk-chasing-dots{top:50%;margin:auto;width:40px;height:40px;position:relative;text-align:center;animation:sk-chasingDotsRotate 2s linear infinite}.sk-chasing-dots .sk-child{width:60%;height:60%;display:inline-block;position:absolute;top:0;border-radius:100%;animation:sk-chasingDotsBounce 2s ease-in-out infinite}.sk-chasing-dots .sk-dot2{top:auto;bottom:0;animation-delay:-1s}@keyframes sk-chasingDotsRotate{to{transform:rotate(1turn)}}@keyframes sk-chasingDotsBounce{0%,to{transform:scale(0)}50%{transform:scale(1)}}"] });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.0.0", ngImport: i0__namespace, type: SkChasingDotsComponent, decorators: [{
                type: i0.Component,
                args: [{
                        selector: 'sk-chasing-dots',
                        templateUrl: './sk-chasing-dots.component.html',
                        styleUrls: ['./sk-chasing-dots.component.scss']
                    }]
            }] });

    var SkCubeGridComponent = /** @class */ (function (_super) {
        __extends(SkCubeGridComponent, _super);
        function SkCubeGridComponent() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        return SkCubeGridComponent;
    }(AbstractLoaderDirective));
    SkCubeGridComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.0.0", ngImport: i0__namespace, type: SkCubeGridComponent, deps: null, target: i0__namespace.ɵɵFactoryTarget.Component });
    SkCubeGridComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.0.0", type: SkCubeGridComponent, selector: "sk-cube-grid", usesInheritance: true, ngImport: i0__namespace, template: "<!--\nCopyright (c) 2015 Tobias Ahlin\n\nPermission is hereby granted, free of charge, to any person obtaining a copy of\nthis software and associated documentation files (the \"Software\"), to deal in\nthe Software without restriction, including without limitation the rights to\nuse, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of\nthe Software, and to permit persons to whom the Software is furnished to do so,\nsubject to the following conditions:\n\nThe above copyright notice and this permission notice shall be included in all\ncopies or substantial portions of the Software.\n\nTHE SOFTWARE IS PROVIDED \"AS IS\", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR\nIMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS\nFOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR\nCOPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER\nIN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN\nCONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.\n-->\n\n<div class=\"sk-cube-grid\" [class.colored]=\"!backgroundColor\">\n    <div class=\"sk-cube sk-cube1\" [style.background-color]='backgroundColor'></div>\n    <div class=\"sk-cube sk-cube2\" [style.background-color]='backgroundColor'></div>\n    <div class=\"sk-cube sk-cube3\" [style.background-color]='backgroundColor'></div>\n    <div class=\"sk-cube sk-cube4\" [style.background-color]='backgroundColor'></div>\n    <div class=\"sk-cube sk-cube5\" [style.background-color]='backgroundColor'></div>\n    <div class=\"sk-cube sk-cube6\" [style.background-color]='backgroundColor'></div>\n    <div class=\"sk-cube sk-cube7\" [style.background-color]='backgroundColor'></div>\n    <div class=\"sk-cube sk-cube8\" [style.background-color]='backgroundColor'></div>\n    <div class=\"sk-cube sk-cube9\" [style.background-color]='backgroundColor'></div>\n</div>\n", styles: [".sk-cube-grid{position:relative;top:50%;width:40px;height:40px;margin:auto}.sk-cube-grid .sk-cube{width:33.33%;height:33.33%;float:left;animation:sk-cubeGridScaleDelay 1.3s ease-in-out infinite}.sk-cube-grid .sk-cube1{animation-delay:.2s}.sk-cube-grid .sk-cube2{animation-delay:.3s}.sk-cube-grid .sk-cube3{animation-delay:.4s}.sk-cube-grid .sk-cube4{animation-delay:.1s}.sk-cube-grid .sk-cube5{animation-delay:.2s}.sk-cube-grid .sk-cube6{animation-delay:.3s}.sk-cube-grid .sk-cube7{animation-delay:0s}.sk-cube-grid .sk-cube8{animation-delay:.1s}.sk-cube-grid .sk-cube9{animation-delay:.2s}@keyframes sk-cubeGridScaleDelay{0%,70%,to{transform:scaleX(1)}35%{transform:scale3D(0,0,1)}}"] });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.0.0", ngImport: i0__namespace, type: SkCubeGridComponent, decorators: [{
                type: i0.Component,
                args: [{
                        selector: 'sk-cube-grid',
                        templateUrl: './sk-cube-grid.component.html',
                        styleUrls: ['./sk-cube-grid.component.scss']
                    }]
            }] });

    var SkDoubleBounceComponent = /** @class */ (function (_super) {
        __extends(SkDoubleBounceComponent, _super);
        function SkDoubleBounceComponent() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        return SkDoubleBounceComponent;
    }(AbstractLoaderDirective));
    SkDoubleBounceComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.0.0", ngImport: i0__namespace, type: SkDoubleBounceComponent, deps: null, target: i0__namespace.ɵɵFactoryTarget.Component });
    SkDoubleBounceComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.0.0", type: SkDoubleBounceComponent, selector: "sk-double-bounce", usesInheritance: true, ngImport: i0__namespace, template: "<!--\nCopyright (c) 2015 Tobias Ahlin\n\nPermission is hereby granted, free of charge, to any person obtaining a copy of\nthis software and associated documentation files (the \"Software\"), to deal in\nthe Software without restriction, including without limitation the rights to\nuse, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of\nthe Software, and to permit persons to whom the Software is furnished to do so,\nsubject to the following conditions:\n\nThe above copyright notice and this permission notice shall be included in all\ncopies or substantial portions of the Software.\n\nTHE SOFTWARE IS PROVIDED \"AS IS\", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR\nIMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS\nFOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR\nCOPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER\nIN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN\nCONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.\n-->\n\n<div class=\"sk-double-bounce\" [class.colored]=\"!backgroundColor\">\n    <div class=\"sk-child sk-double-bounce1\" [style.background-color]='backgroundColor'></div>\n    <div class=\"sk-child sk-double-bounce2\" [style.background-color]='backgroundColor'></div>\n</div>\n", styles: [".sk-double-bounce{top:50%;width:40px;height:40px;position:relative;margin:auto}.sk-double-bounce .sk-child{width:100%;height:100%;border-radius:50%;background-color:#333;opacity:.6;position:absolute;top:0;left:0;animation:sk-doubleBounce 2s ease-in-out infinite}.sk-double-bounce .sk-double-bounce2{animation-delay:-1s}@keyframes sk-doubleBounce{0%,to{transform:scale(0)}50%{transform:scale(1)}}"] });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.0.0", ngImport: i0__namespace, type: SkDoubleBounceComponent, decorators: [{
                type: i0.Component,
                args: [{
                        selector: 'sk-double-bounce',
                        templateUrl: './sk-double-bounce.component.html',
                        styleUrls: ['./sk-double-bounce.component.scss']
                    }]
            }] });

    var SkRotatingPlaneComponent = /** @class */ (function (_super) {
        __extends(SkRotatingPlaneComponent, _super);
        function SkRotatingPlaneComponent() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        return SkRotatingPlaneComponent;
    }(AbstractLoaderDirective));
    SkRotatingPlaneComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.0.0", ngImport: i0__namespace, type: SkRotatingPlaneComponent, deps: null, target: i0__namespace.ɵɵFactoryTarget.Component });
    SkRotatingPlaneComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.0.0", type: SkRotatingPlaneComponent, selector: "sk-rotating-plane", usesInheritance: true, ngImport: i0__namespace, template: "<!--\nCopyright (c) 2015 Tobias Ahlin\n\nPermission is hereby granted, free of charge, to any person obtaining a copy of\nthis software and associated documentation files (the \"Software\"), to deal in\nthe Software without restriction, including without limitation the rights to\nuse, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of\nthe Software, and to permit persons to whom the Software is furnished to do so,\nsubject to the following conditions:\n\nThe above copyright notice and this permission notice shall be included in all\ncopies or substantial portions of the Software.\n\nTHE SOFTWARE IS PROVIDED \"AS IS\", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR\nIMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS\nFOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR\nCOPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER\nIN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN\nCONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.\n-->\n\n<div class=\"sk-rotating-plane colored-parent\" [style.background-color]='backgroundColor'></div>\n", styles: [".sk-rotating-plane{position:relative;top:50%;width:40px;height:40px;margin:auto;animation:sk-rotatePlane 1.2s ease-in-out infinite}@keyframes sk-rotatePlane{0%{transform:perspective(120px) rotateX(0deg) rotateY(0deg)}50%{transform:perspective(120px) rotateX(-180.1deg) rotateY(0deg)}to{transform:perspective(120px) rotateX(-180deg) rotateY(-179.9deg)}}"] });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.0.0", ngImport: i0__namespace, type: SkRotatingPlaneComponent, decorators: [{
                type: i0.Component,
                args: [{
                        selector: 'sk-rotating-plane',
                        templateUrl: './sk-rotating-plane.component.html',
                        styleUrls: ['./sk-rotating-plane.component.scss']
                    }]
            }] });

    var SkSpinnerPulseComponent = /** @class */ (function (_super) {
        __extends(SkSpinnerPulseComponent, _super);
        function SkSpinnerPulseComponent() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        return SkSpinnerPulseComponent;
    }(AbstractLoaderDirective));
    SkSpinnerPulseComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.0.0", ngImport: i0__namespace, type: SkSpinnerPulseComponent, deps: null, target: i0__namespace.ɵɵFactoryTarget.Component });
    SkSpinnerPulseComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.0.0", type: SkSpinnerPulseComponent, selector: "sk-spinner-pulse", usesInheritance: true, ngImport: i0__namespace, template: "<!--\nCopyright (c) 2015 Tobias Ahlin\n\nPermission is hereby granted, free of charge, to any person obtaining a copy of\nthis software and associated documentation files (the \"Software\"), to deal in\nthe Software without restriction, including without limitation the rights to\nuse, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of\nthe Software, and to permit persons to whom the Software is furnished to do so,\nsubject to the following conditions:\n\nThe above copyright notice and this permission notice shall be included in all\ncopies or substantial portions of the Software.\n\nTHE SOFTWARE IS PROVIDED \"AS IS\", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR\nIMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS\nFOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR\nCOPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER\nIN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN\nCONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.\n-->\n\n<div class=\"sk-spinner sk-spinner-pulse colored-parent\" [style.background-color]='backgroundColor'></div>\n", styles: [".sk-spinner-pulse{position:relative;top:50%;width:40px;height:40px;margin:auto;border-radius:100%;animation:sk-pulseScaleOut 1s ease-in-out infinite}@keyframes sk-pulseScaleOut{0%{transform:scale(0)}to{transform:scale(1);opacity:0}}"] });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.0.0", ngImport: i0__namespace, type: SkSpinnerPulseComponent, decorators: [{
                type: i0.Component,
                args: [{
                        selector: 'sk-spinner-pulse',
                        templateUrl: './sk-spinner-pulse.component.html',
                        styleUrls: ['./sk-spinner-pulse.component.scss']
                    }]
            }] });

    var SkThreeBounceComponent = /** @class */ (function (_super) {
        __extends(SkThreeBounceComponent, _super);
        function SkThreeBounceComponent() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        return SkThreeBounceComponent;
    }(AbstractLoaderDirective));
    SkThreeBounceComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.0.0", ngImport: i0__namespace, type: SkThreeBounceComponent, deps: null, target: i0__namespace.ɵɵFactoryTarget.Component });
    SkThreeBounceComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.0.0", type: SkThreeBounceComponent, selector: "sk-three-bounce", usesInheritance: true, ngImport: i0__namespace, template: "<!--\nCopyright (c) 2015 Tobias Ahlin\n\nPermission is hereby granted, free of charge, to any person obtaining a copy of\nthis software and associated documentation files (the \"Software\"), to deal in\nthe Software without restriction, including without limitation the rights to\nuse, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of\nthe Software, and to permit persons to whom the Software is furnished to do so,\nsubject to the following conditions:\n\nThe above copyright notice and this permission notice shall be included in all\ncopies or substantial portions of the Software.\n\nTHE SOFTWARE IS PROVIDED \"AS IS\", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR\nIMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS\nFOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR\nCOPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER\nIN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN\nCONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.\n-->\n\n<div class=\"sk-three-bounce\" [class.colored]=\"!backgroundColor\">\n    <div class=\"sk-child sk-bounce1\" [style.background-color]='backgroundColor'></div>\n    <div class=\"sk-child sk-bounce2\" [style.background-color]='backgroundColor'></div>\n    <div class=\"sk-child sk-bounce3\" [style.background-color]='backgroundColor'></div>\n</div>\n", styles: [".sk-three-bounce{top:50%;position:relative;margin:auto;width:80px;text-align:center}.sk-three-bounce .sk-child{width:20px;height:20px;border-radius:100%;display:inline-block;animation:sk-three-bounce 1.4s ease-in-out 0s infinite both}.sk-three-bounce .sk-bounce1{animation-delay:-.32s}.sk-three-bounce .sk-bounce2{animation-delay:-.16s}@keyframes sk-three-bounce{0%,80%,to{transform:scale(0)}40%{transform:scale(1)}}"] });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.0.0", ngImport: i0__namespace, type: SkThreeBounceComponent, decorators: [{
                type: i0.Component,
                args: [{
                        selector: 'sk-three-bounce',
                        templateUrl: './sk-three-bounce.component.html',
                        styleUrls: ['./sk-three-bounce.component.scss']
                    }]
            }] });

    var SkWanderingCubesComponent = /** @class */ (function (_super) {
        __extends(SkWanderingCubesComponent, _super);
        function SkWanderingCubesComponent() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        return SkWanderingCubesComponent;
    }(AbstractLoaderDirective));
    SkWanderingCubesComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.0.0", ngImport: i0__namespace, type: SkWanderingCubesComponent, deps: null, target: i0__namespace.ɵɵFactoryTarget.Component });
    SkWanderingCubesComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.0.0", type: SkWanderingCubesComponent, selector: "sk-wandering-cubes", usesInheritance: true, ngImport: i0__namespace, template: "<!--\nCopyright (c) 2015 Tobias Ahlin\n\nPermission is hereby granted, free of charge, to any person obtaining a copy of\nthis software and associated documentation files (the \"Software\"), to deal in\nthe Software without restriction, including without limitation the rights to\nuse, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of\nthe Software, and to permit persons to whom the Software is furnished to do so,\nsubject to the following conditions:\n\nThe above copyright notice and this permission notice shall be included in all\ncopies or substantial portions of the Software.\n\nTHE SOFTWARE IS PROVIDED \"AS IS\", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR\nIMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS\nFOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR\nCOPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER\nIN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN\nCONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.\n-->\n\n<div class=\"sk-wandering-cubes\" [class.colored]=\"!backgroundColor\">\n    <div class=\"sk-cube sk-cube1\" [style.background-color]='backgroundColor'></div>\n    <div class=\"sk-cube sk-cube2\" [style.background-color]='backgroundColor'></div>\n</div>\n", styles: [".sk-wandering-cubes{top:50%;margin:auto;width:40px;height:40px;position:relative}.sk-wandering-cubes .sk-cube{width:10px;height:10px;position:absolute;top:0;left:0;animation:sk-wanderingCube 1.8s ease-in-out -1.8s infinite both}.sk-wandering-cubes .sk-cube2{animation-delay:-.9s}@keyframes sk-wanderingCube{0%{transform:rotate(0deg)}25%{transform:translateX(30px) rotate(-90deg) scale(.5)}50%{transform:translateX(30px) translateY(30px) rotate(-179deg)}50.1%{transform:translateX(30px) translateY(30px) rotate(-180deg)}75%{transform:translateX(0) translateY(30px) rotate(-270deg) scale(.5)}to{transform:rotate(-1turn)}}"] });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.0.0", ngImport: i0__namespace, type: SkWanderingCubesComponent, decorators: [{
                type: i0.Component,
                args: [{
                        selector: 'sk-wandering-cubes',
                        templateUrl: './sk-wandering-cubes.component.html',
                        styleUrls: ['./sk-wandering-cubes.component.scss']
                    }]
            }] });

    var SkWaveComponent = /** @class */ (function (_super) {
        __extends(SkWaveComponent, _super);
        function SkWaveComponent() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        return SkWaveComponent;
    }(AbstractLoaderDirective));
    SkWaveComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.0.0", ngImport: i0__namespace, type: SkWaveComponent, deps: null, target: i0__namespace.ɵɵFactoryTarget.Component });
    SkWaveComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.0.0", type: SkWaveComponent, selector: "sk-wave", usesInheritance: true, ngImport: i0__namespace, template: "<!--\nCopyright (c) 2015 Tobias Ahlin\n\nPermission is hereby granted, free of charge, to any person obtaining a copy of\nthis software and associated documentation files (the \"Software\"), to deal in\nthe Software without restriction, including without limitation the rights to\nuse, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of\nthe Software, and to permit persons to whom the Software is furnished to do so,\nsubject to the following conditions:\n\nThe above copyright notice and this permission notice shall be included in all\ncopies or substantial portions of the Software.\n\nTHE SOFTWARE IS PROVIDED \"AS IS\", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR\nIMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS\nFOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR\nCOPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER\nIN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN\nCONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.\n-->\n\n<div class=\"sk-wave\" [class.colored]=\"!backgroundColor\">\n    <div class=\"sk-rect sk-rect1\" [style.background-color]='backgroundColor'></div>\n    <div class=\"sk-rect sk-rect2\" [style.background-color]='backgroundColor'></div>\n    <div class=\"sk-rect sk-rect3\" [style.background-color]='backgroundColor'></div>\n    <div class=\"sk-rect sk-rect4\" [style.background-color]='backgroundColor'></div>\n    <div class=\"sk-rect sk-rect5\" [style.background-color]='backgroundColor'></div>\n</div>\n", styles: [".sk-wave{position:relative;top:50%;margin:auto;width:50px;height:40px;text-align:center;font-size:10px}.sk-wave .sk-rect{float:left;margin-right:1px;height:100%;width:6px;display:inline-block;animation:sk-waveStretchDelay 1.2s ease-in-out infinite}.sk-wave .sk-rect1{animation-delay:-1.2s}.sk-wave .sk-rect2{animation-delay:-1.1s}.sk-wave .sk-rect3{animation-delay:-1s}.sk-wave .sk-rect4{animation-delay:-.9s}.sk-wave .sk-rect5{animation-delay:-.8s}@keyframes sk-waveStretchDelay{0%,40%,to{transform:scaleY(.4)}20%{transform:scaleY(1)}}"] });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.0.0", ngImport: i0__namespace, type: SkWaveComponent, decorators: [{
                type: i0.Component,
                args: [{
                        selector: 'sk-wave',
                        templateUrl: './sk-wave.component.html',
                        styleUrls: ['./sk-wave.component.scss']
                    }]
            }] });

    /*
     * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
     * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
     * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
     * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
     * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
     * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
     */
    var Spinkit = {
        skChasingDots: 'sk-chasing-dots',
        skCubeGrid: 'sk-cube-grid',
        skDoubleBounce: 'sk-double-bounce',
        skRotatingPlane: 'sk-rotationg-plane',
        skSpinnerPulse: 'sk-spinner-pulse',
        skThreeBounce: 'sk-three-bounce',
        skWanderingCubes: 'sk-wandering-cubes',
        skWave: 'sk-wave'
    };
    var SPINKIT_COMPONENTS = [
        SkCubeGridComponent,
        SkChasingDotsComponent,
        SkDoubleBounceComponent,
        SkRotatingPlaneComponent,
        SkSpinnerPulseComponent,
        SkThreeBounceComponent,
        SkWanderingCubesComponent,
        SkWaveComponent,
    ];

    /*
     * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
     * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
     * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
     * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
     * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
     * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
     */
    var PendingRequestsInterceptor = /** @class */ (function () {
        function PendingRequestsInterceptor() {
            this._pendingRequests = 0;
            this._pendingRequestsStatus$ = new rxjs.ReplaySubject(1);
            this._filteredUrlPatterns = [];
            this._filteredMethods = [];
            this._filteredHeaders = [];
            this._forceByPass = false;
        }
        Object.defineProperty(PendingRequestsInterceptor.prototype, "pendingRequestsStatus$", {
            get: function () {
                return this._pendingRequestsStatus$.asObservable();
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(PendingRequestsInterceptor.prototype, "pendingRequests", {
            get: function () {
                return this._pendingRequests;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(PendingRequestsInterceptor.prototype, "filteredUrlPatterns", {
            get: function () {
                return this._filteredUrlPatterns;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(PendingRequestsInterceptor.prototype, "filteredMethods", {
            set: function (httpMethods) {
                this._filteredMethods = httpMethods;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(PendingRequestsInterceptor.prototype, "filteredHeaders", {
            set: function (value) {
                this._filteredHeaders = value;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(PendingRequestsInterceptor.prototype, "forceByPass", {
            set: function (value) {
                this._forceByPass = value;
            },
            enumerable: false,
            configurable: true
        });
        PendingRequestsInterceptor.prototype.shouldBypassUrl = function (url) {
            return this._filteredUrlPatterns.some(function (e) {
                return e.test(url);
            });
        };
        PendingRequestsInterceptor.prototype.shouldBypassMethod = function (req) {
            return this._filteredMethods.some(function (e) {
                return e.toUpperCase() === req.method.toUpperCase();
            });
        };
        PendingRequestsInterceptor.prototype.shouldBypassHeader = function (req) {
            return this._filteredHeaders.some(function (e) {
                return req.headers.has(e);
            });
        };
        PendingRequestsInterceptor.prototype.shouldBypass = function (req) {
            return this._forceByPass
                || this.shouldBypassUrl(req.urlWithParams)
                || this.shouldBypassMethod(req)
                || this.shouldBypassHeader(req);
        };
        PendingRequestsInterceptor.prototype.intercept = function (req, next) {
            var _this = this;
            var shouldBypass = this.shouldBypass(req);
            if (!shouldBypass) {
                this._pendingRequests++;
                if (1 === this._pendingRequests) {
                    this._pendingRequestsStatus$.next(true);
                }
            }
            return next.handle(req).pipe(operators.finalize(function () {
                if (!shouldBypass) {
                    _this._pendingRequests--;
                    if (0 === _this._pendingRequests) {
                        _this._pendingRequestsStatus$.next(false);
                    }
                }
            }));
        };
        return PendingRequestsInterceptor;
    }());
    PendingRequestsInterceptor.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.0.0", ngImport: i0__namespace, type: PendingRequestsInterceptor, deps: [], target: i0__namespace.ɵɵFactoryTarget.Injectable });
    PendingRequestsInterceptor.ɵprov = i0__namespace.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "12.0.0", ngImport: i0__namespace, type: PendingRequestsInterceptor, providedIn: 'root' });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.0.0", ngImport: i0__namespace, type: PendingRequestsInterceptor, decorators: [{
                type: i0.Injectable,
                args: [{
                        providedIn: 'root'
                    }]
            }] });
    var PendingRequestsInterceptorProvider = [{
            provide: http.HTTP_INTERCEPTORS,
            useExisting: PendingRequestsInterceptor,
            multi: true
        }];

    /*
     * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
     * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
     * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
     * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
     * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
     * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
     */
    var SpinnerVisibilityService = /** @class */ (function () {
        function SpinnerVisibilityService(pendingRequestsInterceptor) {
            this.pendingRequestsInterceptor = pendingRequestsInterceptor;
            this._visibility$ = new rxjs.ReplaySubject(1);
        }
        Object.defineProperty(SpinnerVisibilityService.prototype, "visibility$", {
            get: function () {
                return this._visibility$.asObservable();
            },
            enumerable: false,
            configurable: true
        });
        SpinnerVisibilityService.prototype.show = function () {
            this.pendingRequestsInterceptor.forceByPass = true;
            this._visibility$.next(true);
        };
        SpinnerVisibilityService.prototype.hide = function () {
            this._visibility$.next(false);
            this.pendingRequestsInterceptor.forceByPass = false;
        };
        return SpinnerVisibilityService;
    }());
    SpinnerVisibilityService.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.0.0", ngImport: i0__namespace, type: SpinnerVisibilityService, deps: [{ token: PendingRequestsInterceptor }], target: i0__namespace.ɵɵFactoryTarget.Injectable });
    SpinnerVisibilityService.ɵprov = i0__namespace.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "12.0.0", ngImport: i0__namespace, type: SpinnerVisibilityService, providedIn: 'root' });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.0.0", ngImport: i0__namespace, type: SpinnerVisibilityService, decorators: [{
                type: i0.Injectable,
                args: [{
                        providedIn: 'root'
                    }]
            }], ctorParameters: function () { return [{ type: PendingRequestsInterceptor }]; } });

    var NgHttpLoaderComponent = /** @class */ (function () {
        function NgHttpLoaderComponent(pendingRequestsInterceptor, spinnerVisibility) {
            this.pendingRequestsInterceptor = pendingRequestsInterceptor;
            this.spinnerVisibility = spinnerVisibility;
            this.spinkit = Spinkit;
            this.visibleUntil = Date.now();
            this.backdrop = true;
            this.debounceDelay = 0;
            this.entryComponent = null;
            this.extraDuration = 0;
            this.filteredHeaders = [];
            this.filteredMethods = [];
            this.filteredUrlPatterns = [];
            this.minDuration = 0;
            this.opacity = '.7';
            this.backdropBackgroundColor = '#f1f1f1';
            this.spinner = Spinkit.skWave;
        }
        NgHttpLoaderComponent.prototype.ngOnInit = function () {
            this.initIsvisibleObservable();
            this.nullifySpinnerIfEntryComponentIsDefined();
            this.initFilters();
        };
        NgHttpLoaderComponent.prototype.initIsvisibleObservable = function () {
            var _this = this;
            var _a = __read(rxjs.partition(this.pendingRequestsInterceptor.pendingRequestsStatus$, function (h) { return h; }), 2), showSpinner$ = _a[0], hideSpinner$ = _a[1];
            this.isVisible$ = rxjs.merge(this.pendingRequestsInterceptor.pendingRequestsStatus$
                .pipe(operators.switchMap(function () { return showSpinner$.pipe(operators.debounce(function () { return rxjs.timer(_this.debounceDelay); })); })), showSpinner$
                .pipe(operators.switchMap(function () { return hideSpinner$.pipe(operators.debounce(function () { return _this.getVisibilityTimer$(); })); })), this.spinnerVisibility.visibility$).pipe(operators.distinctUntilChanged(), operators.tap(function (h) { return _this.updateExpirationDelay(h); }));
        };
        NgHttpLoaderComponent.prototype.nullifySpinnerIfEntryComponentIsDefined = function () {
            if (this.entryComponent) {
                this.spinner = null;
            }
        };
        NgHttpLoaderComponent.prototype.initFilters = function () {
            this.initFilteredUrlPatterns();
            this.initFilteredMethods();
            this.initFilteredHeaders();
        };
        NgHttpLoaderComponent.prototype.initFilteredUrlPatterns = function () {
            var _this = this;
            if (!!this.filteredUrlPatterns.length) {
                this.filteredUrlPatterns.forEach(function (e) { return _this.pendingRequestsInterceptor.filteredUrlPatterns.push(new RegExp(e)); });
            }
        };
        NgHttpLoaderComponent.prototype.initFilteredMethods = function () {
            this.pendingRequestsInterceptor.filteredMethods = this.filteredMethods;
        };
        NgHttpLoaderComponent.prototype.initFilteredHeaders = function () {
            this.pendingRequestsInterceptor.filteredHeaders = this.filteredHeaders;
        };
        NgHttpLoaderComponent.prototype.updateExpirationDelay = function (showSpinner) {
            if (showSpinner) {
                this.visibleUntil = Date.now() + this.minDuration;
            }
        };
        NgHttpLoaderComponent.prototype.getVisibilityTimer$ = function () {
            return rxjs.timer(Math.max(this.extraDuration, this.visibleUntil - Date.now()));
        };
        return NgHttpLoaderComponent;
    }());
    NgHttpLoaderComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.0.0", ngImport: i0__namespace, type: NgHttpLoaderComponent, deps: [{ token: PendingRequestsInterceptor }, { token: SpinnerVisibilityService }], target: i0__namespace.ɵɵFactoryTarget.Component });
    NgHttpLoaderComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.0.0", type: NgHttpLoaderComponent, selector: "ng-http-loader", inputs: { backdrop: "backdrop", backgroundColor: "backgroundColor", debounceDelay: "debounceDelay", entryComponent: "entryComponent", extraDuration: "extraDuration", filteredHeaders: "filteredHeaders", filteredMethods: "filteredMethods", filteredUrlPatterns: "filteredUrlPatterns", minDuration: "minDuration", opacity: "opacity", backdropBackgroundColor: "backdropBackgroundColor", spinner: "spinner" }, ngImport: i0__namespace, template: "<div id=\"spinner\"\n     *ngIf=\"isVisible$ | async\"\n     [class.backdrop]=\"backdrop\"\n     [style.opacity]=\"opacity\"\n     [ngStyle]=\"{'background-color': backdrop ? backdropBackgroundColor : 'transparent'}\">\n\n    <ng-container *ngComponentOutlet=\"entryComponent\"></ng-container>\n\n    <sk-cube-grid\n        *ngIf=\"spinner === spinkit.skCubeGrid\"\n        [backgroundColor]=\"backgroundColor\">\n    </sk-cube-grid>\n\n    <sk-chasing-dots\n        *ngIf=\"spinner === spinkit.skChasingDots\"\n        [backgroundColor]=\"backgroundColor\">\n    </sk-chasing-dots>\n\n    <sk-double-bounce\n        *ngIf=\"spinner === spinkit.skDoubleBounce\"\n        [backgroundColor]=\"backgroundColor\">\n    </sk-double-bounce>\n\n    <sk-rotating-plane\n        *ngIf=\"spinner === spinkit.skRotatingPlane\"\n        [backgroundColor]=\"backgroundColor\">\n    </sk-rotating-plane>\n\n    <sk-spinner-pulse\n        *ngIf=\"spinner === spinkit.skSpinnerPulse\"\n        [backgroundColor]=\"backgroundColor\">\n    </sk-spinner-pulse>\n\n    <sk-three-bounce\n        *ngIf=\"spinner === spinkit.skThreeBounce\"\n        [backgroundColor]=\"backgroundColor\">\n    </sk-three-bounce>\n\n    <sk-wandering-cubes\n        *ngIf=\"spinner === spinkit.skWanderingCubes\"\n        [backgroundColor]=\"backgroundColor\">\n    </sk-wandering-cubes>\n\n    <sk-wave\n        *ngIf=\"spinner === spinkit.skWave\"\n        [backgroundColor]=\"backgroundColor\">\n    </sk-wave>\n\n</div>\n\n", styles: ["#spinner{top:50%;left:50%;transform:translate(-50%,-50%);position:fixed;z-index:9999}#spinner.backdrop{top:0;left:0;height:100%;width:100%;display:flex;align-items:center;justify-content:center;transform:none}::ng-deep .colored-parent,::ng-deep .colored>div{background-color:#333}"], components: [{ type: SkCubeGridComponent, selector: "sk-cube-grid" }, { type: SkChasingDotsComponent, selector: "sk-chasing-dots" }, { type: SkDoubleBounceComponent, selector: "sk-double-bounce" }, { type: SkRotatingPlaneComponent, selector: "sk-rotating-plane" }, { type: SkSpinnerPulseComponent, selector: "sk-spinner-pulse" }, { type: SkThreeBounceComponent, selector: "sk-three-bounce" }, { type: SkWanderingCubesComponent, selector: "sk-wandering-cubes" }, { type: SkWaveComponent, selector: "sk-wave" }], directives: [{ type: i11__namespace.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }, { type: i11__namespace.NgStyle, selector: "[ngStyle]", inputs: ["ngStyle"] }, { type: i11__namespace.NgComponentOutlet, selector: "[ngComponentOutlet]", inputs: ["ngComponentOutlet", "ngComponentOutletInjector", "ngComponentOutletContent", "ngComponentOutletNgModuleFactory"] }], pipes: { "async": i11__namespace.AsyncPipe } });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.0.0", ngImport: i0__namespace, type: NgHttpLoaderComponent, decorators: [{
                type: i0.Component,
                args: [{
                        selector: 'ng-http-loader',
                        templateUrl: './ng-http-loader.component.html',
                        styleUrls: ['./ng-http-loader.component.scss']
                    }]
            }], ctorParameters: function () { return [{ type: PendingRequestsInterceptor }, { type: SpinnerVisibilityService }]; }, propDecorators: { backdrop: [{
                    type: i0.Input
                }], backgroundColor: [{
                    type: i0.Input
                }], debounceDelay: [{
                    type: i0.Input
                }], entryComponent: [{
                    type: i0.Input
                }], extraDuration: [{
                    type: i0.Input
                }], filteredHeaders: [{
                    type: i0.Input
                }], filteredMethods: [{
                    type: i0.Input
                }], filteredUrlPatterns: [{
                    type: i0.Input
                }], minDuration: [{
                    type: i0.Input
                }], opacity: [{
                    type: i0.Input
                }], backdropBackgroundColor: [{
                    type: i0.Input
                }], spinner: [{
                    type: i0.Input
                }] } });

    var NgHttpLoaderModule = /** @class */ (function () {
        function NgHttpLoaderModule() {
        }
        NgHttpLoaderModule.forRoot = function () {
            return {
                ngModule: NgHttpLoaderModule,
                providers: [
                    PendingRequestsInterceptorProvider,
                ]
            };
        };
        return NgHttpLoaderModule;
    }());
    NgHttpLoaderModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.0.0", ngImport: i0__namespace, type: NgHttpLoaderModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    NgHttpLoaderModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.0.0", ngImport: i0__namespace, type: NgHttpLoaderModule, declarations: [NgHttpLoaderComponent, SkCubeGridComponent, SkChasingDotsComponent, SkDoubleBounceComponent, SkRotatingPlaneComponent, SkSpinnerPulseComponent, SkThreeBounceComponent, SkWanderingCubesComponent, SkWaveComponent], imports: [i11.CommonModule], exports: [NgHttpLoaderComponent, SkCubeGridComponent, SkChasingDotsComponent, SkDoubleBounceComponent, SkRotatingPlaneComponent, SkSpinnerPulseComponent, SkThreeBounceComponent, SkWanderingCubesComponent, SkWaveComponent] });
    NgHttpLoaderModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.0.0", ngImport: i0__namespace, type: NgHttpLoaderModule, imports: [[
                i11.CommonModule,
            ]] });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.0.0", ngImport: i0__namespace, type: NgHttpLoaderModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        declarations: __spreadArray([
                            NgHttpLoaderComponent
                        ], __read(SPINKIT_COMPONENTS)),
                        imports: [
                            i11.CommonModule,
                        ],
                        exports: __spreadArray([
                            NgHttpLoaderComponent
                        ], __read(SPINKIT_COMPONENTS))
                    }]
            }] });

    /*
     * Public API Surface of ng-http-loader
     */

    /**
     * Generated bundle index. Do not edit.
     */

    exports.AbstractLoaderDirective = AbstractLoaderDirective;
    exports.NgHttpLoaderComponent = NgHttpLoaderComponent;
    exports.NgHttpLoaderModule = NgHttpLoaderModule;
    exports.PendingRequestsInterceptor = PendingRequestsInterceptor;
    exports.PendingRequestsInterceptorProvider = PendingRequestsInterceptorProvider;
    exports.SPINKIT_COMPONENTS = SPINKIT_COMPONENTS;
    exports.SkChasingDotsComponent = SkChasingDotsComponent;
    exports.SkCubeGridComponent = SkCubeGridComponent;
    exports.SkDoubleBounceComponent = SkDoubleBounceComponent;
    exports.SkRotatingPlaneComponent = SkRotatingPlaneComponent;
    exports.SkSpinnerPulseComponent = SkSpinnerPulseComponent;
    exports.SkThreeBounceComponent = SkThreeBounceComponent;
    exports.SkWanderingCubesComponent = SkWanderingCubesComponent;
    exports.SkWaveComponent = SkWaveComponent;
    exports.Spinkit = Spinkit;
    exports.SpinnerVisibilityService = SpinnerVisibilityService;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
